src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,655 @@
+
+                ReadOnlySequence<char> remainingChars = chars;
+                int originalBytesLength = bytes.Length;
+                Span<byte> remainingBytes = bytes;
+                Encoder encoder = encoding!.GetEncoder();
+                bool isFinalSegment;
+
+                do
+                {
+                    remainingChars.GetFirstSpan(out ReadOnlySpan<char> firstSpan, out SequencePosition next);
+                    isFinalSegment = remainingChars.IsSingleSegment;
+
+                    int bytesWrittenJustNow = encoder.GetBytes(firstSpan, remainingBytes, flush: isFinalSegment);
+                    remainingBytes = remainingBytes.Slice(bytesWrittenJustNow);
+                    remainingChars = remainingChars.Slice(next);
+                } while (!isFinalSegment);
+
+                return originalBytesLength - remainingBytes.Length; // total number of bytes we wrote
+            }
+        }
+
+        /// <summary>
+        /// Encodes the specified <see cref="ReadOnlySequence{Char}"/> into a <see cref="byte"/> array using the specified <see cref="Encoding"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="chars"/> should be encoded.</param>
+        /// <param name="chars">The <see cref="ReadOnlySequence{Char}"/> to encode to <see langword="byte"/>s.</param>
+        /// <returns>A <see cref="byte"/> array which represents the encoded contents of <paramref name="chars"/>.</returns>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static byte[] GetBytes(this Encoding encoding, in ReadOnlySequence<char> chars)
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (chars.IsSingleSegment)
+            {
+                // If the incoming sequence is single-segment, one-shot this.
+
+                ReadOnlySpan<char> span = chars.FirstSpan;
+
+                byte[] retVal = new byte[encoding!.GetByteCount(span)];
+                int actualBytesWritten = encoding.GetBytes(span, retVal);
+
+                Debug.Assert(actualBytesWritten == retVal.Length, "Unexpected number of elements written.");
+
+                return retVal;
+            }
+            else
+            {
+                // If the incoming sequence is multi-segment, create a stateful Encoder
+                // and use it as the workhorse. On the final iteration we'll pass flush=true.
+
+                Encoder encoder = encoding!.GetEncoder();
+
+                // Maintain a list of all the segments we'll need to concat together.
+                // These will be released back to the pool at the end of the method.
```

Comment:
Was looking at ways how to do this. Creating an array pool-backed `IBufferWriter<T>`, even for internal use, makes me uneasy because there's a chance that these extension methods may become normal virtual methods in a future release. And I don't want to risk passing such a dangerous `IBufferWriter<T>` instance into user code.

Another avenue I explored was refactoring this out into a helper routine, but then that helper routine's signature becomes:

```cs
static TResult MyHelper<T, U, Transcoder, Result>(Transcoder, SomeDelegate countOutput, SomeOtherDelegate transcodeOutput, YetAnotherDelegate resultFactory)
```

The end result is a bunch of indirection, which harms maintainability, and which doesn't actually save many lines of code.

I'll hold off on this for now unless it becomes a problem. But thanks for the suggestion!